### PR TITLE
PoC interpreting evaluator

### PR DESCRIPTION
**Pull request by jpschorr**
PoC basic (though anemic) value type, logical plan (needs to evolve to a graph model), and sketch of an interpreting evaluator (FROM, FROM_AT, UNPIVOT, WHERE, and SELECT)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-eval/src/env.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+pub trait Bindings<T> {
+    fn get(&self, name: &BindingsName) -> Option<&T>;
+}
+
+impl Bindings<Value> for Tuple {
+    fn get(&self, name: &BindingsName) -> Option<&Value> {
+        match name {
+            BindingsName::CaseSensitive(s) => self.0.get(s),
+            BindingsName::CaseInsensitive(s) => {
+                //TODO
```
Comment (am357) [inline partiql-eval/src/env.rs:15]: Other than the fact that we should perhaps treat the case insensitive differently, what do you expect to be as part of `TODO`?

**Review by am357 — APPROVED** [review]
Since this is a PoC that enables us to kick-start the work, I don't see a point in reviewing it in details; I had a cursory review and approving it order to get the code in asap.

File: partiql-eval/src/env.rs (hunk 1 above)
Comment (jpschorr) [reply partiql-eval/src/env.rs:15]: That's the todo. It's not actually doing case sensitivity currently.

File: partiql-eval/src/eval.rs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+    fn push_tuple(&mut self, bindings: Tuple, ctx: &dyn EvalContext) {
+        let is_new = self.seen.insert(bindings.clone());
+        if is_new {
+            self.output.push_tuple(bindings, ctx)
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct EvalOutputAccumulator {
```
Comment (alancai98) [inline partiql-eval/src/eval.rs:442]: Might be a naive question but what is `EvalOutputAccumulator` for? Also, why does is it's only field a `Bag`? Couldn't there be other output types emitted by PartiQL?

File: partiql-eval/src/plan.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                expr,
+                as_key,
+                at_key: _,
+                out,
+            }) => Box::new(eval::EvalFrom::new(
+                self.plan_values(expr),
+                &as_key,
+                self.plan_bindings(*out),
+            )),
+            _ => panic!("Unevaluable bexpr"),
```
Comment (alancai98) [inline partiql-eval/src/plan.rs:33]: To check my understanding, only `BindingsExpr::From` is currently supported/implemented in the POC's evaluator planner?

**Review by alancai98 — APPROVED** [review]
Also gave a general review of this POC PR (focusing primarily on the `partiql-eval` crate. Left a couple comments to check my understanding. Further questions about the finer details may come up as we continue the implementation of the logical interpreter.

File: partiql-eval/src/eval.rs (hunk 2 above)
Comment (jpschorr) [reply partiql-eval/src/eval.rs:442]: The PoC only models Bindings -> Bindings relations properly.

We will need to build out Bindings -> Value (i.e., Select Value) and more properly model Value -> Bindings (i.e., From, Unpivot) .

Yes, a Value -> Value expression (e.g. `1+2`) at the head will return a single value, not a bag, but is not modelled currently.

**Pull request merged**